docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,193 @@
+# RFC 003: Taxonomy of potential performance issues in Tendermint 
+
+## Changelog
+
+- 2021-09-02: Created initial draft (@wbanfield)
+
+## Abstract
+
+This document discusses the various sources of performance issue in Tendermint and
+attempts to clarify what work may be required to understand and address them.
+
+## Background
+
+Performance, loosely defined as the ability of a software process to perform its work
+quickly and efficiently under load and within reasonable resource limits, is a frequent
+topic of discussion in the Tendermint project.
+To effectively address any issues with Tendermint performance we need to
+categorize the various issues, understand their potential sources, and gauge their
+impact on users.
+
+Categorizing the different known performance issues will allow us to discuss and fix them
+more systematically. This document proposes a rough taxonomy of performance issues
+and highlights areas where more research into potential performance problems is required.
+
+Understanding Tendermint's performance limitations will also be critically important
+as we make changes to many of its subsystems. Performance is a central concern for
+upcoming decisions regarding the `p2p` protocol, RPC message encoding and structure,
+database usage and selection, and consensus protocol updates.
+
+### References
+
+* [ADR 57](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-057-RPC.md): 
+Discussion of alternative RPC frameworks for Tendermint.
+* [Issue 1319](https://github.com/tendermint/tendermint/issues/1319): More context 
+on signature aggregation.
+* [Issue 2187](https://github.com/tendermint/tendermint/issues/2187): More context
+on discussions of faster hashing algorithms in Tendermint.
+* [ABCI Commit description](https://github.com/tendermint/spec/blob/master/spec/abci/apps.md#commit):
+More context on locking the mempool during commit.
+* [ABCI local client code](https://github.com/tendermint/tendermint/blob/511bd3eb7f037855a793a27ff4c53c12f085b570/abci/client/local_client.go#L84):
+Clear look at lack of serialization in the local client.
+
+## Discussion
+
+This section attempts to delineate the different sections of Tendermint functionality
+that are often cited as having performance issues. It raises questions and suggests
+lines of inquiry that may be valuable for better understanding Tendermint's performance issues.
+
+As a note: We should avoid quickly adding many microbenchmarks or package level benchmarks. 
```

Comment:
yes for overall performance of a node this adds little value, but for individual packages I would say it still makes sense to afdd benchmarks in order to know when regressions occur without having to spin up networks.

Reply:
Package-level benchmarks can help, but it probably doesn't make sense to spend a lot of time writing them preemptively. The first place to add performance testing IMO is in the form of system metrics (e.g., Prometheus latency distributions), which will give you insight into where the network spends most of its time. Once we have that, it makes sense to drill down into individual packages that get a lot of load, and file down runtime and allocation issues. 

If we had infinite time and resources, of course, we could write benchmarks for everything, but I don't think it makes sense to do it exhaustively before we have some higher-level perf data.